File: JavaScript/516. Longest Palindromic Subsequence.js
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindromeSubseq = function(s) {
    var str="";
    var n=s.length;
    for (var i = n - 1;  i >= 0; i--) {
        str += s[i] + "";
    }
    var dp = Array(n+1).fill().map(v=>Array(n+1).fill(0));
    for(var i = n-1; i >= 0; i--){
        for(var j = n-1; j>=0; j--){
            if(s[i] == str[j]){
                dp[i][j] = dp[i+1][j+1]+1;
            } 
            else{
                dp[i][j] = Math.max(dp[i][j+1], dp[i+1][j]);
            }
        }
    }
    return dp[0][0];
};

File: C++/516. Longest Palindromic Subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        string str=s;
        reverse(str.begin(),str.end()); 
        int n=s.size();
        vector<vector<int>>dp(n+1,vector<int>(n+1,-1));
        return helper(dp,s,str,n-1,n-1);
    }
    int helper(vector<vector<int>>&dp,string &s,string &str,int i,int j){
        if(i<0 || j<0){
            return 0;
        }
        if(dp[i][j]!=-1){return dp[i][j];}
        if(s[i]==str[j]){
            return dp[i][j]=1+helper(dp,s,str,i-1,j-1);
        }
        return dp[i][j]=max(helper(dp,s,str,i-1,j),helper(dp,s,str,i,j-1));
    }
};
